Change conf path XDG_CACHE_HOME to XDG_STATE_HOME

jNullj opened the pull request:
Keepassxc saves application state at XDG_CACHE_HOME which can be cleared on some systems periodicly.
This is not desireable as app state like window size is not consistent when openning the app.
To avoid this, this PR commit is switching the path to XDG_STATE_HOME which is more fitting based on the [freedesktop basedir spec](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html), this will allow to prevent state file deletion as well.

Changes made based on how qt implements QStandardPaths (very similar to internal qt code)

Also added code to migrate old config files to new location and cleanup files from old location.

Fixes #9738 

[NOTE]: # ( Describe your changes in detail, why is this change required? )
[NOTE]: # ( Explain large or complex code modifications. )
[NOTE]: # ( If it fixes an open issue, please add "Fixes #XXX" )


## Screenshots
[TIP]:  # ( Do not include screenshots of your actual database! )
N/A

## Testing strategy
[NOTE]: # ( Please describe in detail how you tested your changes. )
[TIP]:  # ( We expect new code to be covered by unit tests and documented with doc blocks! )

Manually tested file changes in `~/.local/state` and that migration from `~/.cache` works.

## Type of change
[NOTE]: # ( Please remove all lines which don't apply. )
- ✅ New feature (change that adds functionality)
- ✅ Breaking change (causes existing functionality to change)
- ✅ Documentation (non-code change)

jNullj committed:
Change conf path XDG_CACHE_HOME to XDG_STATE_HOME

Keepassxc saves application state at XDG_CACHE_HOME which can be cleared on some systems periodicly.
This is not desireable as app state like window size is not consistent when openning the app.
To avoid this this commit is switching the path to XDG_STATE_HOME which is more fitting based on the freedesktop basedir spec (https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html), this will allow to prevent state file deletion as well.

Solves #9738

jNullj committed:
update docs for switch to XDG_STATE_HOME

phoerious reviewed (changes requested):
You should migrate existing settings from the old location.

jNullj committed:
fix format

jNullj committed:
Migrate old conf file to new location on linux

phoerious commented on src/core/Config.cpp:
```diff
@@ -472,6 +472,26 @@ void Config::init(const QString& configFileName, const QString& localConfigFileN
+    // Upgrade from previous KeePassXC version which stores its config
+    // in ~/.cache on Linux instead of ~/.local/state.
+    // Move file to correct location before continuing.
+    QString oldLocalConfigPath;
+    oldLocalConfigPath = QStandardPaths::writableLocation(QStandardPaths::GenericCacheLocation) + "/keepassxc";
+    QString suffix;
+#ifdef QT_DEBUG
+    suffix = "_debug";
+#endif
+    oldLocalConfigPath += QString("/keepassxc%1.ini").arg(suffix);
+    oldLocalConfigPath = QDir::toNativeSeparators(oldLocalConfigPath);
+    if (!localConfigFileName.isEmpty() && !QFile::exists(localConfigFileName) && QFile::exists(oldLocalConfigPath)) {
```
I don't think you should check for localConfigFileName being empty. That's just something that shouldn't happen and would be an error elsewhere. I think you should just check if it exists and if not, then check the old location. The definition of oldLocalConfigPath can be moved into the smaller scope of the corresponding if statement.

jNullj replied:
I missed that, the isEmpty there is not relevant, removed it with https://github.com/keepassxreboot/keepassxc/pull/9755/commits/2eecf9892eed4c73376fc76415a2ff43979e5e03
I also understand that this is not optimal to generate oldLocalConfigPath on every startup, i changed the conditions and moved it inside the first if statement, we will only generated if the config is missing, fixed in https://github.com/keepassxreboot/keepassxc/pull/9755/commits/d77ef9d4f71759d216f2ca9e04c1db3fdb038dae

phoerious commented on src/core/Config.cpp:
```diff
@@ -512,7 +532,17 @@ QPair<QString, QString> Config::defaultConfigFiles()
     configPath = QStandardPaths::writableLocation(QStandardPaths::GenericConfigLocation) + "/keepassxc";
-    localConfigPath = QStandardPaths::writableLocation(QStandardPaths::GenericCacheLocation) + "/keepassxc";
+    // Qt does not support XDG_STATE_HOME yet, change this once XDG_STATE_HOME is added
+    QString xdgStateHome;
+    xdgStateHome = QFile::decodeName(qgetenv("XDG_STATE_HOME"));
```
Combine these two.

jNullj replied:
My bad, fixed with https://github.com/keepassxreboot/keepassxc/pull/9755/commits/0cd5ce425e1e26854413914b0fc4e088372e5b4e

jNullj committed:
remove unnecessary condition

jNullj committed:
Improve code readability

jNullj committed:
Improve performance

Only generate oldLocalConfigPath if the local config is missing and the old config path should be checked.
Before this commit oldLocalConfigPath would always get generated, while it won't be needed most of the time.

phoerious commented on src/core/Config.cpp:
```diff
@@ -476,19 +476,21 @@ void Config::init(const QString& configFileName, const QString& localConfigFileN
-    oldLocalConfigPath = QStandardPaths::writableLocation(QStandardPaths::GenericCacheLocation) + "/keepassxc";
-    QString suffix;
+    if (!QFile::exists(localConfigFileName)) {
+        QString oldLocalConfigPath;
+        oldLocalConfigPath = QStandardPaths::writableLocation(QStandardPaths::GenericCacheLocation) + "/keepassxc";
```
You should not separate declaration/definition and assignment. Rest looks good.

jNullj replied:
Fixed in https://github.com/keepassxreboot/keepassxc/pull/9755/commits/85a2b740ffd2423a30d90dd9e31e2d77ac3f54aa

jNullj committed:
Improve code readability

jNullj committed:
Merge branch 'develop' into feat/switch-to-XDG_STATE_HOME-9738

jNullj committed:
Merge branch 'develop' into feat/switch-to-XDG_STATE_HOME-9738

phoerious commented:
Looks good now. Thanks for your patience.
